### ISSUE TITLE
Questions about compiling on windows

### ISSUE THREAD

**Issue by user234683** [EionRobb/purple-discord#201]
I'm new to C. I have some modifications I want to make, so I've been trying for a few days to get it to compile on windows. At first I tried to do the cflags and stuff manually since the makefile has a lot of hardcoded directories and version numbers (didn't know at the time where to find the exact versions and I figured it would be easier than messing with the directory structure). It seemed everything was in place except for the `-lpurple` library link. I have all the other library files in msys and I can locate the exact files used by the `-l` switch, but I cannot figure out what file this one is supposed to refer to. I tried to link to libpurple.dll in `C:\Program Files (x86)\Pidgin` even though it was quite a long shot. Surprisingly it compiled, but unsurprisingly it didn't work. Says it can't find the `purple_init_plugin` symbol. I even downloaded the exact pidgin 2.0.11 tarball but can't find anything except `purple.h`, and if I'm understanding correctly you can't link to header files. Before I spend time setting up the exact directory structure, version numbers, and whatnot, I'm wondering where is this library I'm supposed to link to?

Another question. I tried to follow [these directions](https://developer.pidgin.im/wiki/BuildingWinPidgin) to set up the build environment properly. One of the things the makefile assumes and the directions say is to set up mingw-4.7.2. But even the first file, ​`binutils-2.24` no longer exists; even going to the sourceforge page manually, only the last 3 versions are available. Is it important that the mingw version is 4.7.2? Or can I just use the latest versions of these packages?

**Comment by EionRobb** [comment]
I just clicked on the binutils-2.24 link (just under the "Download the following:" heading on the https://developer.pidgin.im/wiki/BuildingWinPidgin page, and it downloaded from sourceforge after a short delay

Any gcc command starting with `-l` means "link a library".  On windows this'll mean `lib{whatever}.dll`.  So `-lpurple` means `libpurple.dll`.

If it says it can't find `purple_init_plugin` then it means `PURPLE_PLUGINS` isn't defined, which is what the the `-DPURPLE_PLUGINS` flag does in the Makefile.

The directory structure that it's expecting should be the directory structure needed to compile Pidgin.  Once you've compiled Pidgin then you should have a good setup to be able to compile the plugin, including all the right headers and dll's

Sounds like your custom Makefile is pretty close to working though!  Just that `-DPURPLE_PLUGINS` should do the trick :)

**Comment by user234683** [comment]
> I just clicked on the binutils-2.24 link (just under the "Download the following:" heading on the https://developer.pidgin.im/wiki/BuildingWinPidgin page, and it downloaded from sourceforge after a short delay

Whoops I meant the dependencies for pidgin 2.10.11 at https://developer.pidgin.im/wiki/BuildingWinPidgin?version=161, since the makefile specifies that version (is it important that it's that version?) . I linked the latest one by mistake. 

>If it says it can't find purple_init_plugin then it means PURPLE_PLUGINS isn't defined, which is what the the -DPURPLE_PLUGINS flag does in the Makefile.

It turns out that wasn't the issue. PURPLE_PLUGINS will get defined regardless because in `purple_compat.h`:
```
#ifndef PURPLE_PLUGINS
#define PURPLE_PLUGINS
#endif
```
The issue was that in my big compile command, I forgot to actually put the file I was compiling - `libdiscord.c` - and gcc just made a big file full of dependencies without any warnings. Now the error I get is ``(13:23:14) plugins: C:\Program Files (x86)\Pidgin\plugins\libdiscord.dll is not loadable: `C:\Program Files (x86)\Pidgin\plugins\libdiscord.dll': The specified procedure could not be found.``

After digging through the source code for pidgin and glib, I deduced that this error occurs when attempting to load the module with `LoadLibraryW`, not because of libpurple/glib trying to use `GetProcAddress`. This implies then, that the error is occurring during load-time dynamic linking inside `libdiscord.dll`. I'm not sure if this is because I'm missing an `-l` flag somewhere (would gcc even compile it in that case?) or if it's because I compiled against different versions than used in my installation. Either way, I'm going to try setting up the build environment properly and build against pidgin 2.12.0.

**Comment by user234683** [comment]
I got it to compile and made my changes. A few issues though:
- I followed the instructions given at https://developer.pidgin.im/wiki/BuildingWinPidgin for creating the mingw-4.7.2 folder. However, the gcc in this folder isn't functional, trying to compile something with it  results in no output (no file, no messages, nothing). I found [this question](https://stackoverflow.com/questions/32827417/mingw-gcc-does-nothing) about it. I'll try to figure out what's missing, as I'd like to make a single zip file with the folder structure and all the dependencies required to compile Pidgin and all the instructions completed, as the current process is unreasonable. But I figured I'd mention it in case this has come up before and you know anything about it. I ended up just changing the makefile to point to a mingw installation made with the installer.
- Using the compiled plugin (even without any changes made to it), everything works fine except that when quitting pidgin, it crashes with `Microsoft Visual C++ Runtime Library: This application has requested the Runtime to terminate it in an unusual way.` After closing that dialog, the "This program has stopped working" thing comes up. The problem signature is
```
Problem signature:
  Problem Event Name:	APPCRASH
  Application Name:	pidgin.exe
  Application Version:	2.12.0.0
  Application Timestamp:	58c20b6e
  Fault Module Name:	libgcc_s_dw2-1.dll
  Fault Module Version:	0.0.0.0
  Fault Module Timestamp:	592c8b14
  Exception Code:	40000015
  Exception Offset:	00015697
  OS Version:	6.1.7601.2.1.0.768.3
  Locale ID:	1033
  Additional Information 1:	7a7a
  Additional Information 2:	7a7adb4d0bbbda4629a4be56ea5a5204
  Additional Information 3:	3837
  Additional Information 4:	3837cf6c5ef4506e65cde350f5b0c2ab
```
I'm not sure what `libgcc_s_dw2-1.dll` is doing there. That file isn't even present in the installation. I found [this question](https://stackoverflow.com/questions/4702732/the-program-cant-start-because-libgcc-s-dw2-1-dll-is-missing), but `-static-libgcc` is already being used. Adding `-static-libstdc++` changes nothing. I found [this thread](https://sourceforge.net/p/notepad-plus/discussion/482781/thread/81e3242c/) about the same issue but in a notepad++ plugin. It turned out in that case to be a conflict with another plugin. But removing all other plugins in Pidgin except for mine doesn't fix it like in that case. As before, mentioning this in case it's come up before I start debugging it. If you don't have any ideas about these problems, feel free to close this issue.
